src/meta/src/stream/fragmenter/rewrite/delta_join.rs

```diff
@@ -338,4 +288,132 @@ impl StreamFragmenter {
 
         Ok(union)
     }
+
+    pub(in super::super) fn build_delta_join_without_arrange(
+        &self,
+        state: &mut BuildFragmentGraphState,
+        current_fragment: &mut StreamFragment,
+        mut node: StreamNode,
+    ) -> Result<StreamNode> {
+        match &node.node {
+            Some(Node::DeltaIndexJoin(node)) => node,
+            _ => unreachable!(),
+        };
+
+        let arrange_1 = node.input.remove(1);
+        let arrange_0 = node.input.remove(0);
+
+        // TODO: when distribution key is added to catalog, chain and delta join won't have any
+        // exchange in-between. Then we can safely remove this function.
+        fn check_no_exchange(node: &StreamNode) {
+            if let Node::ExchangeNode(_) = node.get_node().unwrap() {
+                panic!("exchange not allowed between delta join and arrange");
+            }
+        }
+
+        check_no_exchange(&arrange_0);
+        check_no_exchange(&arrange_1);
```

Comment:
Does that mean currently we disallow queries with a plan that has exchange between chain and delta join?

Reply:
Yes, we disallow it for now.

Reply:
If we need to support it, shuffle should come before chain.